doc/md/guides: added article "Partial Indexes in SQLite" (#1223)
ariga/atlas: doc/md/guides/sqlite/partial-indexes.md

```sql
SELECT * FROM vaccination_data;
```
```console title="Output"
+----+--------------------+-------+----------------+------------+
| id |      country       | title |     names      | vaccinated |
+----+--------------------+-------+----------------+------------+
| 1  | Poland             | Er.   | Travis Freeman | No         |
| 2  | Australia          | Mr.   | Hu Dodson      | No         |
| 3  | Vietnam            | Ms.   | Amery Herman   | No         |
| 4  | Peru               | Mr.   | Brynne Mann    | Yes        |
| 5  | Chile              | Er.   | Nora Mitchell  | No         |
| 6  | Brazil             | Er.   | Tanner Oneal   | No         |
| 7  | Vietnam            | Mr.   | Ora Conway     | Yes        |
| 8  | United Kingdom     | Er.   | Quinn Waters   | No         |
| 9  | Russian Federation | Er.   | Xyla Holloway  | No         |
| 10 | Norway             | Mr.   | Macy Sullivan  | No         |
.
.
.
| 576000 | Ukraine   | Dr.   | Kuame Gay         | Yes        |
+--------+-----------+-------+-------------------+------------+
```

:::info
You can also beautify tables in SQLite like shown above, by using the command `.mode table`
:::

> In the following example, suppose we want a list of doctors from India that have taken the vaccine. If we want to use non-partial index, we can create it on the “vaccinated” column with the following command:

Reviewer: ```suggestion
In the following example, suppose we want a list of doctors from India that have taken the vaccine. If we want to use a non-partial index, we can create it on the "vaccinated" column with the following command:
```

```sql
CREATE INDEX 
    vaccinated_idx 
ON 
    vaccination_data(vaccinated);
```

Now, let's check the size of the index that we created, with the following command:

```sql
SELECT NAME, sum(pgsize) AS size FROM dbstat GROUP BY NAME ORDER BY size DESC;
```
```console title="Output"
+------------------+----------+
|       name       |   size   |
+------------------+----------+
| vaccination_data | 22224896 |
| vaccinated_idx   | 6492160  |
| sqlite_schema    | 4096     |
+------------------+----------+
```
Notice that the total size of our index vaccinated_idx is 6492160 bytes (~6 MB).

:::info
The DBSTAT virtual table is a read-only eponymous virtual table that returns information about the amount of disk space used to store the content of an SQLite database. To know more about DBSTAT, visit the official documentation page [here](https://www.sqlite.org/dbstat.html)
:::

Now, suppose we want to accelerate the same query using the partial index. Let's begin by dropping the existing index that we created earlier:

```sql
DROP INDEX vaccinated_idx;
```

> In the following command, we have created an index with a `WHERE` clause that precisely describes list of doctors from India that have taken the vaccine.

Reviewer: ```suggestion
In the following command, we will create an index with a `WHERE` clause that precisely describes the list of doctors from India that have taken the vaccine.
```

```sql
CREATE INDEX 
    vaccinated_idx
ON 
    vaccination_data(vaccinated)
WHERE 
    vaccinated = 'Yes' AND country = 'India' AND title = 'Dr';
```

> Let’s verify if the index we created is being used in the query with `WHERE` clause by the following command:

Reviewer: ```suggestion
Let’s verify if the index we created is being used in the query with a `WHERE` clause by running the following command:
```

```sql
EXPLAIN QUERY PLAN 
SELECT
        *
FROM    
        vaccination_data
WHERE
        vaccinated = 'Yes' AND country = 'India' AND title = 'Dr';
```

```console title="Output"
QUERY PLAN
`--SEARCH vaccination_data USING INDEX vaccinated_idx
```

> We confirmed that the index vaccinated_idx is being used while running the query above. Let's check again the size of the index that we created, with the following command:

Reviewer: ```suggestion
We confirmed that the index vaccinated_idx is being used while running the query above. Let's check the size of the index that we created again, with the following command:
```

Notice that the total size of our index vaccinated_idx is just 4096 bytes. In our example, the index size for the partial index took significantly less space (4 KB) compared to the non-partial index that we created earlier on the 'vaccinated' column (~6 MB).

:::info
Learn more about partial indexes in SQLite from official documentation [here](https://www.sqlite.org/partialindex.html)
:::

We have seen that creating a partial index is a better choice where only a small subset of the values stored in the database are accessed frequently. Now, let's see how we can easily manage partial indexes using Atlas.

### Managing Partial Indexes is easy with Atlas

Managing partial indexes and database schemas in SQLite can be confusing and error-prone. Atlas is an open-source project which allows us to manage our database using a simple and easy-to-understand declarative syntax (similar to Terraform). We will now learn how to manage partial indexes using Atlas.

:::info
If you are just getting started, install the latest version of Atlas using the guide to [setting up Atlas](https://atlasgo.io/cli/getting-started/setting-up).
:::

> #### Managing Partial Index in Atlas

Reviewer: ```suggestion
#### Managing Partial Indexes in Atlas
```

We will first use the `atlas schema inspect` command to get an HCL representation of the table which we created earlier by using the Atlas CLI:

```console
atlas schema inspect -u "sqlite://vaccination_data.db" > schema.hcl
```
```schema title="schema.hcl"
table "vaccination_data" {
  schema = schema.main
  column "id" {
    null = false
    type = integer
  }
  column "country" {
    null    = true
    type    = varchar(100)
  }
  column "title" {
    null    = true
    type    = text
  }
  column "names" {
    null    = true
    type    = varchar(255)
  }
  column "vaccinated" {
    null    = true
    type    = varchar(255)
  }
  primary_key {
    columns = [column.id]
  }
}
schema "main" {
}
```

Now, lets add the following index definition to the file:

```schema
  index "vaccinated_idx" {
    columns = [column.vaccinated]
    where   = "(vaccinated = 'Yes' AND country = 'India' AND title = 'Dr.')"
  }
```

> Save the changes in schema.hcl file and apply the schema changes on the database by using the following command:

Reviewer: ```suggestion
Save the changes in the schema.hcl file and apply the changes on the database by using the following command:
```

```console
atlas schema apply -u "sqlite://vaccination_data.db" -f schema.hcl
```

Atlas generates the necessary SQL statements to add the new partial index to the database schema. Press Enter while the `Apply` option is highlighted to apply the changes:

```console
-- Planned Changes:
-- Create index "vaccinated_idx" to table: "vaccination_data"
CREATE INDEX `vaccinated_idx` ON `vaccination_data` (`vaccinated`) WHERE (vaccinated = 'Yes' AND country = 'India' AND title = 'Dr.')
✔ Apply
  Abort
```

To verify that our new index was created, open the database command line tool and run:

```sql
.index
```
```console title="Output"
vaccinated_idx
```

Amazing! Our new partial index is now created!

> ### Limitation of using Partial Index

Reviewer: ```suggestion
### Limitation of using Partial Indexes
```

> Partial indexes are useful in cases where we know ahead of time that a table is most frequently queried with a certain WHERE clause. As applications evolve, access patterns to the database also change. Consequently, we may find ourselves in a situation where our index no longer covers many queries, causing them to become resource consuming and slow.

Reviewer: ```suggestion
Partial indexes are useful in cases where we know ahead of time that a table is most frequently queried with a certain WHERE clause. As applications evolve, access patterns to the database also change. Consequently, we may find ourselves in a situation where our index no longer covers many queries, causing them to become resource-consuming and slow.
```
